src/DotNetLightning.Core/Crypto/Keys.fs

```diff
@@ -0,0 +1,584 @@
+    
+[<AutoOpen>]
+module NBitcoinArithmethicExtensions =
+    /// The functions in this module may fail but it does not return Result and just throw Exception in
+    /// case of failure. Why? because the error case is very, very unlikely to happen.
+    /// (e.g. two keys we operate were complement of each other.)
+    /// It is interesting to consider if it is possible for the attacker to intentionally make this happen,
+    /// but it is way beyond my skills, so we just ignore for now.
+
+    let Secp256k1 = CryptoUtils.impl.newSecp256k1()
+
+    type Key with
+        static member Mul(lhs: Key, rhs: Key): Key =
+            let lhsBytes = lhs.ToBytes()
+            let rhsBytes = rhs.ToBytes()
+            let tweak = ReadOnlySpan(lhsBytes)
+            match Secp256k1.PrivateKeyTweakMultiply(tweak, rhsBytes.AsSpan()) with
+            | true -> Key rhsBytes
+            | false -> failwith "failed to multiply Keys"
+
+        static member Add(lhs: Key, rhs: Key): Key =
+            let lhsBytes = lhs.ToBytes()
+            let rhsBytes = rhs.ToBytes()
+            let tweak = ReadOnlySpan(lhsBytes)
+            match Secp256k1.PrivateKeyTweakAdd(tweak, rhsBytes.AsSpan()) with
+            | true -> Key rhsBytes
+            | false -> failwithf "failed to add Keys"
+
+    type PubKey with
+        member this.ExpandToBytes(): array<byte> =
+            let compressedPubKeyBytes = this.ToBytes()
+            let compressedPubKeyBytesSpan = ReadOnlySpan(compressedPubKeyBytes)
+            match Secp256k1.PublicKeyParse(compressedPubKeyBytesSpan) with
+            | true, expandedPubKeyBytes -> expandedPubKeyBytes
+            | false, _ -> failwithf "Failed  to expand public key %A" this
+
+        static member CompressFromBytes(expandedPubKeyBytes: array<byte>): PubKey =
+            let expandedPubKeyBytesSpan = ReadOnlySpan(expandedPubKeyBytes)
+            match Secp256k1.PublicKeySerializeCompressed(expandedPubKeyBytesSpan) with
+            | true, compressedPubKeyBytes -> PubKey compressedPubKeyBytes
+            | false, _ -> failwith "Failed to compress pubkey"
+
+        static member Mul(pubKey: PubKey, key: Key): PubKey =
+            let keyBytes = key.ToBytes()
+            let pubKeyBytes = pubKey.ExpandToBytes()
+            let tweak = ReadOnlySpan(keyBytes)
+            match Secp256k1.PublicKeyTweakMultiply(tweak, pubKeyBytes.AsSpan()) with
+            | true -> PubKey.CompressFromBytes pubKeyBytes
+            | false -> failwith "failed to multiplying PubKey by Key"
+
+        static member Add(lhs: PubKey, rhs: PubKey): PubKey =
+            let lhsBytes = lhs.ToBytes()
+            let rhsBytes = rhs.ToBytes()
+            match Secp256k1.PublicKeyCombine(lhsBytes.AsSpan(), rhsBytes.AsSpan()) with
+            | true, result -> PubKey.CompressFromBytes result
+            | false, _ -> failwith "failed to add PubKeys"
+
+type [<Struct>] FundingPubKey(pubKey: PubKey) =
```

Comment:
One reason is so that equality works. As reference types, two identical keys won't compare as equal if they're different references.

Also saving one layer of indirection seems like a good thing.

Reply:
then what needs to be done here is have the type be a discriminated union of only one element, and then there is no need to mark it with the `[<struct>]` attribute

Reply:
(or a record, rather than a DU)

Reply:
I agree with @knocte , DU (or record. I prefer DU though) achieves the same.
Using struct is possible in every case. (i.e. DU, or Record, or plain old class.). But it makes harder to reason about the behavior. the only case I think it is feasible is that when the inner value is also a value type and thus we can gain performance by preventing heap-allocation.

Reply:
I would go as far as to say that idiomatic F# only needs records and DUs, types (that end up compiled in the same way as C# classes) should only be used when trying to achieve interoperability with C#

Reply:
I've changed the key types to be single-constructor discriminated unions.